common/scala/src/main/scala/whisk/core/connector/Message.scala

```diff
@@ -122,3 +116,81 @@ object PingMessage extends DefaultJsonProtocol {
   def parse(msg: String) = Try(serdes.read(msg.parseJson))
   implicit val serdes = jsonFormat(PingMessage.apply _, "name")
 }
+
+trait EventMessageBody extends Message {}
```

Comment:
What about adding a `val typeName: String` here, which you define for each message. You can then use that when figuring out the type for the EventMessage without "hardcoding" it as constants in `UserEvents`.

Reply:
Could be a good fit except from the fact that I have moved `Activation` into EventMessage block to prevent it from evaluation. So it became inconvenient accessing its class variables, we would need to create an extra instance for that...

Reply:
You don't necessarily need to inline the block

```scala
send(producer, {
  val activation = Activation(...)
  EventMessage(...)
})
```

Should be fine!